### PR TITLE
EKA-G20 | Introduced parallel API for Link confirmation. Also introduced the /on-confirm API

### PR DESCRIPTION
**Pull request by angshu**

### CODE REVIEW

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/ErrorMap.java  2
- src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java  4
- src/main/java/in/projecteka/consentmanager/link/link/Link.java  1
         

Clones added
============
- src/main/java/in/projecteka/consentmanager/clients/LinkServiceClient.java  1
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5581080&bid=18294712)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/link/link/Link.java  1
- src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java  4
- src/main/java/in/projecteka/consentmanager/clients/ErrorMap.java  2
         

Clones added
============
- src/main/java/in/projecteka/consentmanager/clients/LinkServiceClient.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5581080&bid=18294712)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/ErrorMap.java  2
- src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java  4
- src/main/java/in/projecteka/consentmanager/link/link/Link.java  1
         

Clones added
============
- src/main/java/in/projecteka/consentmanager/clients/LinkServiceClient.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5581080&bid=18294712)

File: src/main/java/in/projecteka/consentmanager/link/link/model/TokenConfirmation.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package in.projecteka.consentmanager.link.link.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@AllArgsConstructor
+@Data
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/link/model/TokenConfirmation.java:7]: Can you use @value annotation?

File: src/main/java/in/projecteka/consentmanager/link/link/Link.java [hunk 2]
```diff
@@ -119,4 +140,91 @@
...
+
+    private String serializeConfirmationFromHIP(LinkConfirmationResult confirmationResult) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            return objectMapper.writeValueAsString(confirmationResult);
+        } catch (JsonProcessingException e) {
+            logger.error("[Link] Can not serialize response from HIP", e);
+        }
+        return null;
+    }
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/link/Link.java:219]: Can you use the Serializer.from and don't change the method to add log, I have done it separate PR.

File: src/main/java/in/projecteka/consentmanager/link/link/Link.java [hunk 3]
```diff
@@ -119,4 +140,91 @@
...
+
+    private Mono<LinkConfirmationResult> deserializeConfirmationFromHIP(String responseBody) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            return Mono.just(objectMapper.readValue(responseBody, LinkConfirmationResult.class));
+        } catch (JsonProcessingException e) {
+            logger.error("[Link] Can not deserialize response from HIP", e);
+        }
+        return Mono.empty();
+    }
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/link/Link.java:229]: In your Next PR, you should use Serializer.tryTo.

File: src/main/java/in/projecteka/consentmanager/link/link/Link.java [hunk 4]
```diff
@@ -23,6 +39,11 @@
     private final LinkServiceClient linkServiceClient;
     private final LinkRepository linkRepository;
     private final CentralRegistry centralRegistry;
+    private final LinkServiceProperties serviceProperties;
+    private final CacheAdapter<String, String> linkResults;
+
+
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/link/Link.java:45]: Nit: Remove extra space.

File: src/main/java/in/projecteka/consentmanager/link/link/LinkController.java [hunk 5]
```diff
@@ -49,4 +53,28 @@
...
     }
+
+    /**
+     * This API is intended for a CM App.
+     * e.g. a mobile app or from other channels
+     * @param linkRefNumber
+     * @param patientLinkRequest
+     * @return
+     */
+    @PostMapping("/v1/links/link/confirm")
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/link/LinkController.java:64]: If we add comments, better explain what's this API all about too.

File: src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+
+    public boolean hasResponseId(){
+        if (resp == null) {
+            return false;
+        }
+        if (resp.getRequestId() == null || "".equals(resp.getRequestId())) {
+            return false;
+        }
+        return true;
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java:34]: return response != null && !StringUtils.isEmpty(resp.getRequestId());

https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/StringUtils.html#StringUtils--

File: src/main/java/in/projecteka/consentmanager/link/link/model/TokenConfirmation.java (hunk 1 above)
Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/link/link/model/TokenConfirmation.java]: Typically Jackson would use reflection and use javabeans properties std (setter) to bind from JSON to class. @Value didn't play well, but seems to work now.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/clients/ErrorMap.java  2
- src/main/java/in/projecteka/consentmanager/link/link/Link.java  1
- src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java  1
         

Complexity decreasing per file
==============================
+ src/main/java/in/projecteka/consentmanager/link/discovery/model/patient/response/DiscoveryResult.java  -3
         

Clones added
============
- src/main/java/in/projecteka/consentmanager/clients/LinkServiceClient.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5581080&bid=18294712)

File: src/main/java/in/projecteka/consentmanager/link/link/Link.java (hunk 2 above)
Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/link/link/Link.java:217]: will add to the next PR.

File: src/main/java/in/projecteka/consentmanager/link/link/LinkController.java (hunk 5 above)
Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/link/link/LinkController.java:64]: I plan to crosslink with @link to the Specs.

File: src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java (hunk 6 above)
Comment (angshu) [reply src/main/java/in/projecteka/consentmanager/link/link/model/LinkConfirmationResult.java]: done

File: src/main/java/in/projecteka/consentmanager/link/link/model/TokenConfirmation.java (hunk 1 above)
Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/link/link/model/TokenConfirmation.java]: Yeah, configuration is added to make it work :)
